### PR TITLE
Remove nil values from arrays when uploading descriptive metadata

### LINKED ISSUE #3470: Spreadsheet upload: Error message for contributor.name.parallelValue.structuredValue.value
`Line 5: #/components/schemas/Contributor does not allow null values for druid:nb406th1306 (2022-04-12 22:47:00 UTC)`
[spreadsheet_upload_test.xlsx](https://github.com/sul-dlss/argo/files/8476998/spreadsheet_upload_test.xlsx)

### PR DESCRIPTION
**Pull request by aaron-collier**
## Why was this change made? 🤔

Closes #3470 

This addresses the issue where if a descriptive metadata upload spreadsheet as a record with blank values in `contributor1.*` for instance, but does have values in `contributor3.*` that creates parameters with nil values such as:

```
{
  :purl=>"https://purl.stanford.edu/jr825qh8124",
  :title=>[{:value=>"foo bar 2"}],
  :contributor=>[
    nil,
    nil,
    {:name=>[
      nil,
      nil, 
      {:parallelValue=>[
        nil,
        nil, 
        {:structuredValue=>[
          {:value=>"Parallel 1 part 1", :type=>"name"},
          {:value=>"1800-1900", :type=>"life dates"}
        ]},
        {:structuredValue=>[
          {:value=>"Parallel 2 part 1", :type=>"name"},
          {:value=>"marchioness", :type=>"term of address"}]}]}]}]
}
```

This recursively removes the nils from the parameters resulting in:

```
{
  :purl=>"https://purl.stanford.edu/jr825qh8124",
  :title=>[{:value=>"foo bar 2"}],
  :contributor=>[
    {:name=>[
      {:parallelValue=>[
        {:structuredValue=>[
          {:value=>"Parallel 1 part 1", :type=>"name"},
          {:value=>"1800-1900", :type=>"life dates"}
        ]},
        {:structuredValue=>[
          {:value=>"Parallel 2 part 1", :type=>"name"},
          {:value=>"marchioness", :type=>"term of address"}]}]}]}]
}
```

And avoiding `#/components/schemas/Contributor does not allow null values` errors.

## How was this change tested? 🤨

⚡ ⚠ If this change has cross service impact (including writing to shared file systems or interacting with other SDR APIs), ***run [integration tests](https://github.com/sul-dlss/infrastructure-integration-test)*** and/or test in [stage|qa] environment, in addition to specs. ⚡

⚡ ⚠ If this change updates the Argo UI, run all integration tests that use Argo and create a PR on the integration test repo to fix anything this change breaks. ⚡

### CODE REVIEW

File: app/services/description_import.rb [hunk 1]
```diff
@@ -68,4 +68,10 @@ def key?(what, key)
       what.key?(key)
     end
   end
+
+  def compact_params(params)
+    params.each do |key, value|
```
Comment (jcoyne) [inline app/services/description_import.rb:73]: Does this only work at the top level?  Is that good enough or do we need to ensure this happens at any level of the tree?

Comment (aaron-collier) [reply app/services/description_import.rb:73]: Good catch. Not sure how likely that is, but this small change makes it recursive and I've updated the description above.

**Review by jcoyne — CHANGES_REQUESTED** [review]
Please add a test for this.

**Comment by aaron-collier** [comment]
@jcoyne test added.

File: app/services/description_import.rb [hunk 2]
```diff
@@ -71,7 +71,10 @@ def key?(what, key)
 
   def compact_params(params)
     params.each do |key, value|
-      params[key].compact! if value.is_a?(Array)
+      next unless value.is_a?(Array)
```
Comment (jcoyne) [inline app/services/description_import.rb:74]: Won't this kill the recursion? I'm not sure the test shows that this works on nested arrays that have nils.

**Review by jcoyne — CHANGES_REQUESTED** [review]

File: app/services/description_import.rb (hunk 2 above)
Comment (jcoyne) [reply app/services/description_import.rb:74]: Can you test with the data you display in the PR?

Comment (aaron-collier) [reply app/services/description_import.rb:74]: i can make that change. Just to point out, this does show the functionality. If I remove the compact_params method and run this test I get a "Cannot insert nil..." error - so I believe this is properly exercising the code, but I can make deeper test data, sure.

Comment (aaron-collier) [reply app/services/description_import.rb:74]: @jcoyne ok, this is now testing with the data from the comment above.

**Review by jcoyne — APPROVED** [review]

**Pull request merged**